overlord/snapstate/snapstate.go

```diff
@@ -2217,36 +2217,124 @@ func LinkNewBaseOrKernel(st *state.State, name string) (*state.TaskSet, error) {
 	linkSnap := st.NewTask("link-snap", fmt.Sprintf(i18n.G("Make snap %q (%s) available to the system during remodel"), snapsup.InstanceName(), snapst.Current))
 	linkSnap.Set("snap-setup-task", prepareSnap.ID())
 	linkSnap.WaitFor(prepareSnap)
-
 	// we need this for remodel
 	ts := state.NewTaskSet(prepareSnap, linkSnap)
 	ts.MarkEdge(prepareSnap, DownloadAndChecksDoneEdge)
+	if info.Type() == snap.TypeKernel {
+		// kernel snaps can carry boot assets
+		gadgetUpdate := st.NewTask("update-gadget-assets", fmt.Sprintf(i18n.G("Update assets from %s %q (%s) for remodel"), snapsup.Type, snapsup.InstanceName(), snapst.Current))
+		gadgetUpdate.Set("snap-setup-task", prepareSnap.ID())
+		gadgetUpdate.WaitFor(linkSnap)
+		ts.AddTask(gadgetUpdate)
+	}
 	return ts, nil
 }
 
-func AddLinkNewBaseOrKernel(st *state.State, ts *state.TaskSet) (*state.TaskSet, error) {
-	var snapSetupTaskID string
+func findSnapSetupTask(tasks []*state.Task) (*state.Task, *SnapSetup, error) {
 	var snapsup SnapSetup
-	allTasks := ts.Tasks()
-	for _, tsk := range allTasks {
+	for _, tsk := range tasks {
 		if tsk.Has("snap-setup") {
-			snapSetupTaskID = tsk.ID()
 			if err := tsk.Get("snap-setup", &snapsup); err != nil {
-				return nil, err
+				return nil, nil, err
 			}
-			break
+			return tsk, &snapsup, nil
 		}
 	}
-	if snapSetupTaskID == "" {
+	return nil, nil, nil
+}
+
+func AddLinkNewBaseOrKernel(st *state.State, ts *state.TaskSet) (*state.TaskSet, error) {
```

Comment:
the new and old functions should have used/changed here should have doc comments as LinkNewBaseOrKernel does.